src/lib/expression/evaluation/expression_evaluator.cpp

```diff
@@ -0,0 +1,1054 @@
+          case FunctionType::Concatenate: return _evaluate_concatenate(expression.arguments);
+        }
+      } else {
+        Fail("Function can only be evaluated to a string");
+      }
+      // clang-format on
+  }
+}
+
+template <>
+std::shared_ptr<ExpressionResult<std::string>> ExpressionEvaluator::_evaluate_extract_expression<std::string>(
+    const ExtractExpression& extract_expression) {
+  const auto from_result = evaluate_expression_to_result<std::string>(*extract_expression.from());
+
+  switch (extract_expression.datetime_component) {
+    case DatetimeComponent::Year:
+      return _evaluate_extract_substr<0, 4>(*from_result);
+    case DatetimeComponent::Month:
+      return _evaluate_extract_substr<5, 2>(*from_result);
+    case DatetimeComponent::Day:
+      return _evaluate_extract_substr<8, 2>(*from_result);
+
+    case DatetimeComponent::Hour:
+    case DatetimeComponent::Minute:
+    case DatetimeComponent::Second:
+      Fail("Hour, Minute and Second not available in String Datetimes");
+  }
+}
+
+template <typename R>
+std::shared_ptr<ExpressionResult<R>> ExpressionEvaluator::_evaluate_extract_expression(
+    const ExtractExpression& extract_expression) {
+  Fail("Only Strings (YYYY-MM-DD) supported for Dates right now");
+}
+
+template <size_t offset, size_t count>
+std::shared_ptr<ExpressionResult<std::string>> ExpressionEvaluator::_evaluate_extract_substr(
+    const ExpressionResult<std::string>& from_result) {
+  std::shared_ptr<ExpressionResult<std::string>> result;
+
+  std::vector<std::string> values(from_result.size());
+
+  from_result.as_view([&](const auto& from_view) {
+    for (auto chunk_offset = ChunkOffset{0}; chunk_offset < from_view.size(); ++chunk_offset) {
+      if (!from_view.null(chunk_offset)) {
+        DebugAssert(from_view.value(chunk_offset).size() == 10u,
+                    "Invalid DatetimeString '"s + from_view.value(chunk_offset) + "'");
+        values[chunk_offset] = from_view.value(chunk_offset).substr(offset, count);
+      }
+    }
+  });
+
+  return std::make_shared<ExpressionResult<std::string>>(std::move(values), from_result.nulls);
+}
+
+template <typename R>
+std::shared_ptr<ExpressionResult<R>> ExpressionEvaluator::_evaluate_negate_expression(
+    const NegateExpression& negate_expression) {
+  std::vector<R> values;
+  std::vector<bool> nulls;
+
+  _resolve_to_expression_result(*negate_expression.argument(), [&](const auto& argument_result) {
+    using ArgumentType = typename std::decay_t<decltype(argument_result)>::Type;
+
+    // clang-format off
+    if constexpr (!std::is_same_v<ArgumentType, std::string> && std::is_same_v<R, ArgumentType>) {
+      values.resize(argument_result.size());
+      for (auto chunk_offset = ChunkOffset{0}; chunk_offset < argument_result.size(); ++chunk_offset) {
+        values[chunk_offset] = -argument_result.values[chunk_offset];
+      }
+      nulls = argument_result.nulls;
```

Comment:
Have you considered taking argument_results by `&&` and using `std::forward` for cases like this? In most cases, the result of an expression should only be used once, right?

Reply:
Put that on the list of things to eval later, if that is okay with you? https://github.com/hyrise/hyrise/issues/946#issuecomment-403292600